### PR TITLE
Use PostgreSQL

### LINKED ISSUE #17: Configure Databases
**ACCEPTANCE CRITERIA**
- [ ] Use the same database backend in development/test environments as in production.
- [ ] All environments should use Postgres

### PR DESCRIPTION
**Pull request by no-reply**
Closes #17.

This works on local and on travis, but I haven't tried a production deployment, and possibly have broken it in `config/database.yml`

### CODE REVIEW

File: config/database.yml.travis [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+development:
+  <<: *default
+  database: mahonia_development
+  username: mahonia
+  password: password1
+
+test:
+  <<: *default
+  database: travis_ci_test
```
Comment (mark-dce) [inline config/database.yml.travis:15]: I'm not sure why we're not just saying `database: mahonia_test` on line 15 and checking a single `config/database.yml` into source control.  I'm not getting what benefit or flexibility this gives us.  And it seems sort of complex.

File: .travis.yml [hunk 2]
```diff
@@ -7,7 +7,11 @@ jdk:
   - oraclejdk8
 rvm:
   - 2.4.2
+before_script:
+  - cp config/database.yml.travis config/database.yml
+  - psql -c 'create database travis_ci_test;' -U postgres;
```
Comment (mark-dce) [inline .travis.yml:12]: Can't we just say 'mahonia_test' in line 12 and get rid of line 11?

**Review by mark-dce — DISMISSED** [review]
I'm not following the need to have a separate database config file for travis.  Wouldn't it be simpler to just use the same name for the test database regardless of whether we're running test in a local environment or on a CI server?

File: .travis.yml (hunk 2 above)
Comment (no-reply) [reply .travis.yml:12]: No. It's also necessary to create a role that would allow rails to log in and create the database, e.g. https://github.com/curationexperts/mahonia/pull/132/files#diff-04c6e90faac2675aa89e2176d2eec7d8R27. Using `travis_ci_test` uses the role that travis supports out of the box.

I had a small amount of trouble setting up the `mahonia` role and decided it was better to just go with travis's approach. See also: https://docs.travis-ci.com/user/database-setup/#PostgreSQL

File: config/database.yml.travis (hunk 1 above)
Comment (no-reply) [reply config/database.yml.travis:15]: This is Travis's suggested configuration approach. See: https://docs.travis-ci.com/user/database-setup/#PostgreSQL

**Comment by bess** [comment]
+1 for going with the Travis recommended way of doing things.

**Review by bess — APPROVED** [review]

**Comment by mark-dce** [comment]
@bess & @no-reply -1 to doing things the travis recommended way.  It seems really complicated and I haven't seen any other Hydra/Samvera/Hyrax project do it this way.

Is there any reason it couldn't just be this simple:
https://github.com/curationexperts/nurax/blob/master/.travis.yml#L14
```
services:
  - postgresql
```
This has pattern has been working on nurax for two months without issues & it seems way simpler.

**Comment by mark-dce** [comment]
At least, could we give the nurax config a 15 minute review and either:
1) Update nurax to match here
OR
2) Update here to match nurax
So that we have consistency among projects.

**Comment by no-reply** [comment]
@mark-dce It's not clear to me how nurax is authenticating to create the database. Your suggested solution fails as in this branch: https://travis-ci.org/curationexperts/mahonia/builds/302633127#L782

**Comment by no-reply** [comment]
I was able to get [travis building](https://travis-ci.org/curationexperts/mahonia/builds/302637003) in the `nurax` way by removing the authentication username and password from the configuration.

The downside of this is that developers must ensure their current user can create databases in postgres in order to run the development and test envs on their own machines. I.e. there is no uniform developer setup instruction other than "make sure you have the correct postgres permissions".

**Comment by mark-dce** [comment]
I'm not a fan of having multiple database.yml files, can't we just give folks the necessary instructions to set up their dev environment - I think the travis solution presupposes that's what people are doing anyway for the local test environment.  I think the various rake db commands sort of assume that you have local credentials to create a database `psql -c 'create database mahonia_test;' -U postgres;`.  Wouldn't `rake db:drop` and `rake db:setup` both kind of depend on you having local user access to create and delete new databases for test and dev?

**Comment by no-reply** [comment]
>  Wouldn't rake db:drop and rake db:setup both kind of depend on you having local user access to create and delete new databases for test and dev?

Yes, exactly. If there's a configuration with a username and password, then we can bootstrap that automatically; in the case of the original PR, we added instructions to the `README.md` to grant the relevant permissions. Without a specific authentication configuration, the details of how to setup authentication will depend on the developer's own environment. For myself, I had to do:

```sh
$ sudo su - postgres
$ psql
postgres=# create role tjohnson with createdb login;
postgres=# \q
$ exit
$ bundle exec rake db:create
```
With that done, things work okay unless I use a different unix user in development.

In short, removing the username and password configurations makes Travis work by default, but leaves the developer on their own for figuring out how to grant permissions to an appropriate user(s).

**Comment by mark-dce** [comment]
Thanks!  I like #133 much better.  Sorry was such a pain on this one.
